src/Microsoft.Identity.Client/Internal/Requests/ClientCredentialHelper.cs

```diff
@@ -60,14 +60,23 @@ public static bool ValidateClientAssertion(ClientCredentialWrapper clientCredent
                 }
                 else
                 {
-                    if (clientCredential.Assertion == null || clientCredential.ValidTo != 0)
+                    if ((clientCredential.Assertion == null || clientCredential.ValidTo != 0) && String.IsNullOrWhiteSpace(clientCredential.SignedAssertion))
                     {
                         if (!ValidateClientAssertion(clientCredential, endpoints, sendX5C))
                         {
                             logger.Info("Client Assertion does not exist or near expiry.");
-                            var jwtToken = new JsonWebToken(cryptographyManager, clientId, endpoints?.SelfSignedJwtAudience);
+                            JsonWebToken jwtToken;
+                            
+                            if (clientCredential.Certificate.ClaimsToSign != null && clientCredential.Certificate.ClaimsToSign.Count > 0)
```

Comment:
This is a code smell - your ClientAssertion class should have enough abstraction in it so that ClientCredentialHelper doesn't need to access its internal state (ClaimsToSign). 

Also, a simpler way to check if a collection is not empty is .Any() instead .Count (.Count has not changes to enumerate the unerlying collection)

Reply:
The clientAssertion class is gone but I get your point. I can add the abstraction to the clientCredentialWrapper